Refactor qconv to reduce allocations.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42840 Enable input pointer caching in XNNPACK integration.
* #42178 Change quantizer to account for input tensor's memory format.
* #42008 Call qnnpack's conv setup only if input pointer has changed.
* **#42007 Refactor qconv to reduce allocations.**
* #42006 Simple caching allocator for CPU.

Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

kimishpatel committed:
Refactor qconv to reduce allocations.

Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/conv-run.cc:
```diff
@@ -373,60 +350,23 @@ enum pytorch_qnnp_status qnnpackConv(
       const uint32_t cr = pytorch_qnnp_params.q8dw9.cr;
       const size_t group_stride = (groups + (cr - 1)) & -cr;
 
-      if (any_padding) {
```
Can you explain why this block is removed? From 376 to 414.

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:
```diff
@@ -173,45 +118,32 @@ enum pytorch_qnnp_status qnnpackDeConv(
-  const void** indirection_buffer = (const void**)realloc(
-      deconvolution->indirection_buffer, indirection_buffer_size);
-  if (indirection_buffer == NULL) {
+  pytorch_qnnp_status status = pytorch_qnnp_setup_deconvolution2d_nhwc_q8(
```
So again, all of the logic removed to the left is replaced with `pytorch_qnnp_setup_deconvolution2d_nhwc_q8`?

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/test/convolution-operator-tester.h:
```diff
@@ -24,6 +24,79 @@
+  const bool any_padding = (conv_p.padding[0]| conv_p.padding[1]
+      |conv_p.padding[2] | conv_p.padding[3]) != 0;
+
+  if (any_padding) {
```
OK I see, it is moved here.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/include/conv_utils.h:
```diff
@@ -226,12 +226,6 @@ struct conv_param_t {
-      pytorch_qnnp_log_error(
-          "Per channel quantized weights are not supported for XZP kernels");
-      assert("Failed to initialize QNNPACK conv_param_t struct.");
-    }
```
Thanks for the catch. Strange I removed this. Will reinstate.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/conv-run.cc:
```diff
@@ -323,48 +324,24 @@ enum pytorch_qnnp_status qnnpackConv(
+      input_pixel_stride,
+      output,
+      output_pixel_stride,
+      threadpool);
```
Yes.

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]

kimishpatel committed:
Update on "Refactor qconv to reduce allocations."


Summary:
zero buffer and indirection pointers are allocatoed on every iterations.
With this refactor we create op once for qnnpackconv struct and keep
repopulating indirection pointer as necessary.

For deconv moved much of op creation outside so that we can avoid creating and
destroying ops every time.

Test Plan:
CI quantization tests.
deconvolution-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726972](https://our.internmc.facebook.com/intern/diff/D22726972)

[ghstack-poisoned]
